File: src/swift/3_StructuralPatterns/6_Flyweight/Exercise.swift
import Foundation
import XCTest

class Sentence : CustomStringConvertible
{
  var words: [String]
  var tokens = [Int: WordToken]()

  init(_ plainText: String)
  {
    words = plainText.components(separatedBy: " ")
  }

  subscript(index: Int) -> WordToken
  {
    get
    {
      let wt = WordToken()
      tokens[index] = wt
      return tokens[index]!
    }
  }

  var description: String
  {
    var ws = [String]()
    for i in 0..<words.count
    {
      var w = words[i]
      if let item = tokens[i]
      {
        if (item.capitalize)
        {
          w = w.uppercased()
        }
      }
      ws.append(w)
    }
    return ws.joined(separator: " ")
  }

  class WordToken
  {
    var capitalize: Bool = false
    init(){}
    init(capitalize: Bool)
    {
      self.capitalize = capitalize
    }
  }
}

class UMBaseTestCase : XCTestCase {}

//@testable import Test

class Evaluate: UMBaseTestCase
{
  func simpleTest()
  {
    let s = Sentence("alpha beta gamma")
    s[1].capitalize = true
    XCTAssertEqual("alpha BETA gamma", s.description)
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("simpleTest", simpleTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/3_StructuralPatterns/6_Flyweight/exercise.py
import unittest


class Sentence:
    def __init__(self, plain_text):
        self.words = plain_text.split(' ')
        self.tokens = {}

    def __getitem__(self, item):
        wt = self.WordToken()
        self.tokens[item] = wt
        return self.tokens[item]

    class WordToken:
        def __init__(self, capitalize=False):
            self.capitalize = capitalize

    def __str__(self):
        result = []
        for i in range(len(self.words)):
            w = self.words[i]
            if i in self.tokens and self.tokens[i].capitalize:
                w = w.upper()
            result.append(w)
        return ' '.join(result)


class Evaluate(unittest.TestCase):
    def test_exercise(self):
        s = Sentence('alpha beta gamma')
        s[1].capitalize = True
        self.assertEqual(str(s), 'alpha BETA gamma')